File: src/kernels/matvec.cpp
#include "kernels/matvec.h"
#include <immintrin.h>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <algorithm>

// Horizontal sum of __m256 to float
inline float hsum256_ps(__m256 v) {
    __m128 hi = _mm256_extractf128_ps(v, 1);
    __m128 lo = _mm256_castps256_ps128(v);
    __m128 sum = _mm_add_ps(hi, lo);        // [a0+a4, a1+a5, a2+a6, a3+a7]
    sum = _mm_hadd_ps(sum, sum);            // [a0+a4+a1+a5, ...]
    sum = _mm_hadd_ps(sum, sum);            // [a0+a1+a2+a3+a4+a5+a6+a7, ...]
    return _mm_cvtss_f32(sum);
}

template <uint32_t TILE_K, uint32_t TILE_M>
void matvec_fp16_fp32(const fp16_t* W, const float* x, float* y, uint32_t M, uint32_t K) {
    // Loop order: for m → for k (NOT for k → for m)
    // Bu sıra input vector x'in L1 cache'te kalmasını sağlar.
    // K=2560 × 4 bytes = 10KB, L1 = 48KB → x tamamen L1'de kalır.
    for (uint32_t m = 0; m < M; ++m) {
        const fp16_t* W_row = W + static_cast<size_t>(m) * K;

        __m256 acc_vec = _mm256_setzero_ps();
        float acc_scalar = 0.0f;

        uint32_t k = 0;
        for (; k + 8 <= K; k += 8) {
            __m128i w16 = _mm_loadu_si128(
                reinterpret_cast<const __m128i*>(W_row + k));
            __m256 w32 = _mm256_cvtph_ps(w16);
            __m256 xv  = _mm256_loadu_ps(x + k);
            acc_vec = _mm256_fmadd_ps(w32, xv, acc_vec);
        }

        // Scalar tail for remaining elements
        for (; k < K; ++k) {
            acc_scalar += fp16_to_fp32(W_row[k]) * x[k];
        }

        y[m] += hsum256_ps(acc_vec) + acc_scalar;
    }
}

// Explicit instantiations for known tile configs
template void matvec_fp16_fp32<1024, 512>(const fp16_t*, const float*, float*, uint32_t, uint32_t);
template void matvec_fp16_fp32<2048, 1024>(const fp16_t*, const float*, float*, uint32_t, uint32_t);

// Dispatch: pick tile config based on which platform preset we want.
void matvec_dispatch(const fp16_t* W, const float* x, float* y,
                     uint32_t M, uint32_t K, const ModelConfig& config) {
    (void)config;
    matvec_fp16_fp32<Intel13500H_Tiles::TILE_K, Intel13500H_Tiles::TILE_M>(W, x, y, M, K);
}

// --- Fused dequant + matvec for Q8_0 ---
// block_q8_0: 2 bytes d (FP16 scale) + 32 bytes qs (INT8 weights) = 34 bytes
// Each block covers 32 columns of one row
// Row of K columns uses K/32 blocks = K*34/32 bytes
void matvec_q8_0(const void* blocks, const float* x, float* y,
                 uint32_t M, uint32_t K) {
    const uint8_t* base = static_cast<const uint8_t*>(blocks);
    const uint32_t blocks_per_row = K / 32;
    const size_t block_stride = 34; // sizeof(block_q8_0)

    for (uint32_t m = 0; m < M; ++m) {
        float sum = 0.0f;
        const uint8_t* row_ptr = base + static_cast<size_t>(m) * blocks_per_row * block_stride;

        for (uint32_t b = 0; b < blocks_per_row; ++b) {
            const uint8_t* bp = row_ptr + b * block_stride;
            float d = fp16_to_fp32(*reinterpret_cast<const uint16_t*>(bp));
            const int8_t* qs = reinterpret_cast<const int8_t*>(bp + 2);
            const float* x_blk = x + b * 32;

            // Accumulate dot product for this 32-element block
            float block_sum = 0.0f;
            for (int j = 0; j < 32; ++j) {
                block_sum += static_cast<float>(qs[j]) * x_blk[j];
            }
            sum += d * block_sum;
        }
        y[m] += sum;
    }
}

// --- Fused dequant + matvec for Q4_K ---
// block_q4_K: d(FP16) + dmin(FP16) + scales[12] + qs[128] = 144 bytes
// Each block covers 256 columns of one row
void matvec_q4_K(const void* blocks, const float* x, float* y,
                 uint32_t M, uint32_t K) {
    const uint8_t* base = static_cast<const uint8_t*>(blocks);
    const uint32_t blocks_per_row = K / 256;
    const size_t block_stride = 144;

    for (uint32_t m = 0; m < M; ++m) {
        float sum = 0.0f;
        const uint8_t* row_ptr = base + static_cast<size_t>(m) * blocks_per_row * block_stride;

        for (uint32_t b = 0; b < blocks_per_row; ++b) {
            const uint8_t* bp = row_ptr + b * block_stride;
            float d = fp16_to_fp32(*reinterpret_cast<const uint16_t*>(bp));
            float dmin = fp16_to_fp32(*reinterpret_cast<const uint16_t*>(bp + 2));
            const uint8_t* scales = bp + 4;
            const uint8_t* qs = bp + 16;
            const float* x_blk = x + b * 256;

            for (int blk = 0; blk < 4; ++blk) {
                int js = blk * 2;
                uint8_t sc0, mn0, sc1, mn1;
                if (js < 4) {
                    sc0 = scales[js] & 63;
                    mn0 = scales[js + 4] & 63;
                } else {
                    sc0 = (scales[js + 4] & 0x0F) | ((scales[js - 4] >> 6) << 4);
                    mn0 = (scales[js + 4] >> 4)    | ((scales[js] >> 6) << 4);
                }
                if (js + 1 < 4) {
                    sc1 = scales[js + 1] & 63;
                    mn1 = scales[js + 1 + 4] & 63;
                } else {
                    sc1 = (scales[js + 1 + 4] & 0x0F) | ((scales[js + 1 - 4] >> 6) << 4);
                    mn1 = (scales[js + 1 + 4] >> 4)    | ((scales[js + 1] >> 6) << 4);
                }
                float rs0 = d * sc0, rm0 = dmin * mn0;
                float rs1 = d * sc1, rm1 = dmin * mn1;

                int qoff = blk * 32;
                int woff = blk * 64;
                for (int j = 0; j < 16; ++j) {
                    sum += (rs0 * (qs[qoff+j] & 0x0F) - rm0) * x_blk[woff + j];
                    sum += (rs1 * (qs[qoff+16+j] & 0x0F) - rm1) * x_blk[woff + 16 + j];
                    sum += (rs0 * (qs[qoff+j] >> 4) - rm0) * x_blk[woff + 32 + j];
                    sum += (rs1 * (qs[qoff+16+j] >> 4) - rm1) * x_blk[woff + 48 + j];
                }
            }
        }
        y[m] += sum;
    }
}

// --- Fused dequant + matvec for Q2_K ---
// 86 bytes per block, 256 weights per block
void matvec_q2_K(const void* blocks, const float* x, float* y,
                 uint32_t M, uint32_t K) {
    const uint8_t* base = static_cast<const uint8_t*>(blocks);
    const uint32_t blocks_per_row = K / 256;
    const size_t block_stride = 86;

    for (uint32_t m = 0; m < M; ++m) {
        float sum = 0.0f;
        const uint8_t* row_ptr = base + static_cast<size_t>(m) * blocks_per_row * block_stride;

        for (uint32_t b = 0; b < blocks_per_row; ++b) {
            const uint8_t* bp = row_ptr + b * block_stride;
            float d_all = fp16_to_fp32(*reinterpret_cast<const uint16_t*>(bp));
            float m_all = fp16_to_fp32(*reinterpret_cast<const uint16_t*>(bp + 2));
            const uint8_t* scales = bp + 4;
            const uint8_t* qs = bp + 20;
            const float* x_blk = x + b * 256;

            for (int j = 0; j < 256; ++j) {
                int sb = j / 8;
                int s = sb / 2;
                float scale, min_val;
                if (sb % 2 == 0) {
                    scale = d_all * (scales[s] & 0x0F);
                    min_val = m_all * (scales[s] >> 4);
                } else {
                    scale = d_all * (scales[s + 8] & 0x0F);
                    min_val = m_all * (scales[s + 8] >> 4);
                }
                int byte_idx = j / 4;
                int bit_off = (j % 4) * 2;
                int q = (qs[byte_idx] >> bit_off) & 0x03;
                sum += (scale * q - min_val) * x_blk[j];
            }
        }
        y[m] += sum;
    }
}

#ifdef MATVEC_TEST_MAIN
int main() {
    // 4×8 FP16 weight matrix (row-major)
    fp16_t W[4 * 8];
    // 8-length FP32 input vector
    float x[8];
    // 4-length FP32 output vector
    float y[4] = {0};

    // Fill W with simple values: row m has all values = m + 1.0
    for (uint32_t m = 0; m < 4; ++m) {
        for (uint32_t k = 0; k < 8; ++k) {
            W[m * 8 + k] = fp32_to_fp16(static_cast<float>(m + 1));
        }
    }

    // Fill x with values 1..8
    for (uint32_t k = 0; k < 8; ++k) {
        x[k] = static_cast<float>(k + 1);
    }

    matvec_fp16_fp32<8, 4>(W, x, y, 4, 8);

    // Expected:
    // y[0] = 1*(1+2+3+4+5+6+7+8) = 1*36 = 36
    // y[1] = 2*(1+2+3+4+5+6+7+8) = 2*36 = 72
    // y[2] = 3*(1+2+3+4+5+6+7+8) = 3*36 = 108
    // y[3] = 4*(1+2+3+4+5+6+7+8) = 4*36 = 144
    printf("matvec test results:\n");
    for (uint32_t m = 0; m < 4; ++m) {
        printf("  y[%u] = %.1f (expected %.1f) %s\n",
               m, y[m], static_cast<float>(m + 1) * 36.0f,
               y[m] == static_cast<float>(m + 1) * 36.0f ? "OK" : "FAIL");
    }

    // Test with non-tile-aligned K (scalar tail path)
    fp16_t W2[4 * 5];
    float x2[5];
    float y2[4] = {0};

    for (uint32_t m = 0; m < 4; ++m)
        for (uint32_t k = 0; k < 5; ++k)
            W2[m * 5 + k] = fp32_to_fp16(static_cast<float>(m + 1));
    for (uint32_t k = 0; k < 5; ++k)
        x2[k] = static_cast<float>(k + 1);

    matvec_fp16_fp32<8, 4>(W2, x2, y2, 4, 5);

    // Expected: sum(1..5) = 15
    printf("\nTail test (K=5) results:\n");
    bool all_pass = true;
    for (uint32_t m = 0; m < 4; ++m) {
        float expected = static_cast<float>(m + 1) * 15.0f;
        bool ok = (y2[m] == expected);
        if (!ok) all_pass = false;
        printf("  y[%u] = %.1f (expected %.1f) %s\n",
               m, y2[m], expected, ok ? "OK" : "FAIL");
    }
    printf("\n%s\n", all_pass ? "ALL TESTS PASSED" : "SOME TESTS FAILED");

    return all_pass ? 0 : 1;
}
#endif

File: echo-core-zig/src/kernels/matvec.zig
const std = @import("std");
const types = @import("../core/types.zig");
const config = @import("../core/config.zig");
const quant = @import("quant.zig");
const gguf = @import("../gguf/reader.zig");

pub fn matvecFp16Fp32(
    comptime TILE_K: u32,
    comptime TILE_M: u32,
    W: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
) void {
    _ = TILE_K;
    _ = TILE_M;
    const W_fp16: [*]const types.fp16_t = @ptrCast(@alignCast(W));
    var m: u32 = 0;
    while (m < M) : (m += 1) {
        const W_row = W_fp16 + @as(usize, m) * K;
        var acc: f32 = 0;

        var k: u32 = 0;
        while (k < K) : (k += 1) {
            acc += types.fp16_to_fp32(W_row[k]) * x[k];
        }

        y[m] += acc;
    }
}

pub fn matvecF32Fp32(
    comptime TILE_K: u32,
    comptime TILE_M: u32,
    W: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
) void {
    _ = TILE_K;
    _ = TILE_M;
    const W_f32: [*]const f32 = @ptrCast(@alignCast(W));
    var m: u32 = 0;
    while (m < M) : (m += 1) {
        const W_row = W_f32 + @as(usize, m) * K;
        var acc: f32 = 0;

        var k: u32 = 0;
        while (k < K) : (k += 1) {
            acc += W_row[k] * x[k];
        }

        y[m] += acc;
    }
}

pub fn matvecDispatch(
    W: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
    config_: anytype,
) void {
    _ = config_;
    matvecFp16Fp32(config.Intel13500H_Tiles.TILE_K, config.Intel13500H_Tiles.TILE_M, W, x, y, M, K);
}

pub fn matvecDispatchQuant(
    comptime TILE_K: u32,
    comptime TILE_M: u32,
    W: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
    dtype: gguf.GGMLType,
) void {
    switch (dtype) {
        .f16 => matvecFp16Fp32(TILE_K, TILE_M, W, x, y, M, K),
        .f32 => matvecF32Fp32(TILE_K, TILE_M, W, x, y, M, K),
        .q8_0 => matvecQ80(W, x, y, M, K),
        .q6_k => {
            // Q6_K not yet optimized - treat as fp16 for now
            std.debug.print("WARN: Q6_K not optimized, treating as FP16 (may produce incorrect results)\n", .{});
            matvecFp16Fp32(TILE_K, TILE_M, W, x, y, M, K);
        },
        .q4_k => matvecQ4K(W, x, y, M, K),
        .q2_k => matvecQ2K(W, x, y, M, K),
        .iq2_xs, .iq4_xs => {
            // IQ types not yet optimized - treat as fp16 for now
            std.debug.print("WARN: {s} not optimized, treating as FP16 (may produce incorrect results)\n", .{@tagName(dtype)});
            matvecFp16Fp32(TILE_K, TILE_M, W, x, y, M, K);
        },
        else => {
            std.debug.print("WARN: unsupported dtype {s}, falling back to FP16 (may produce incorrect results)\n", .{@tagName(dtype)});
            matvecFp16Fp32(TILE_K, TILE_M, W, x, y, M, K);
        },
    }
}

pub fn matvecQ80(
    blocks: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
) void {
    const blocks_per_row = K / 32;
    const block_stride = 34;

    var m: u32 = 0;
    while (m < M) : (m += 1) {
        var sum: f32 = 0;
        const row_ptr = blocks + @as(usize, m) * blocks_per_row * block_stride;

        var b: u32 = 0;
        while (b < blocks_per_row) : (b += 1) {
            const bp = row_ptr + b * block_stride;
            const d = types.fp16_to_fp32(std.mem.readInt(u16, bp[0..2], .little));
            const qs = @as([*]const i8, @ptrCast(bp + 2));
            const x_blk = x + b * 32;

            var block_sum: f32 = 0;
            var j: u32 = 0;
            while (j < 32) : (j += 1) {
                block_sum += @as(f32, @floatFromInt(qs[j])) * x_blk[j];
            }
            sum += d * block_sum;
        }
        y[m] += sum;
    }
}

pub fn matvecQ4K(
    blocks: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
) void {
    const blocks_per_row = K / 256;
    const block_stride = 144;

    var m: u32 = 0;
    while (m < M) : (m += 1) {
        var sum: f32 = 0;
        const row_ptr = blocks + @as(usize, m) * blocks_per_row * block_stride;

        var b: u32 = 0;
        while (b < blocks_per_row) : (b += 1) {
            const bp = row_ptr + b * block_stride;
            const d = types.fp16_to_fp32(std.mem.readInt(u16, bp[0..2], .little));
            const dmin = types.fp16_to_fp32(std.mem.readInt(u16, bp[2..4], .little));
            const scales = bp[4..];
            const qs = bp[16..];
            const x_blk = x + b * 256;

            var blk: u32 = 0;
            while (blk < 4) : (blk += 1) {
                const js = blk * 2;
                const sc0: u8 = if (js < 4)
                    scales[js] & 63
                else
                    (scales[js + 4] & 0x0F) | ((scales[js - 4] >> 6) << 4);
                const mn0: u8 = if (js < 4)
                    scales[js + 4] & 63
                else
                    (scales[js + 4] >> 4) | ((scales[js] >> 6) << 4);
                const sc1: u8 = if (js + 1 < 4)
                    scales[js + 1] & 63
                else
                    (scales[js + 1 + 4] & 0x0F) | ((scales[js + 1 - 4] >> 6) << 4);
                const mn1: u8 = if (js + 1 < 4)
                    scales[js + 1 + 4] & 63
                else
                    (scales[js + 1 + 4] >> 4) | ((scales[js + 1] >> 6) << 4);

                const rs0 = d * @as(f32, @floatFromInt(sc0));
                const rm0 = dmin * @as(f32, @floatFromInt(mn0));
                const rs1 = d * @as(f32, @floatFromInt(sc1));
                const rm1 = dmin * @as(f32, @floatFromInt(mn1));

                const qoff = @as(usize, blk) * 32;
                const woff = @as(usize, blk) * 64;

                var j: u32 = 0;
                while (j < 16) : (j += 1) {
                    sum += (rs0 * @as(f32, @floatFromInt(qs[qoff + j] & 0x0F)) - rm0) * x_blk[woff + j];
                    sum += (rs1 * @as(f32, @floatFromInt(qs[qoff + 16 + j] & 0x0F)) - rm1) * x_blk[woff + 16 + j];
                    sum += (rs0 * @as(f32, @floatFromInt(qs[qoff + j] >> 4)) - rm0) * x_blk[woff + 32 + j];
                    sum += (rs1 * @as(f32, @floatFromInt(qs[qoff + 16 + j] >> 4)) - rm1) * x_blk[woff + 48 + j];
                }
            }
        }
        y[m] += sum;
    }
}

pub fn matvecQ2K(
    blocks: [*]const u8,
    x: [*]const f32,
    y: [*]f32,
    M: u32,
    K: u32,
) void {
    const blocks_per_row = K / 256;
    const block_stride = 84;

    var m: u32 = 0;
    while (m < M) : (m += 1) {
        var sum: f32 = 0;
        const row_ptr = blocks + @as(usize, m) * blocks_per_row * block_stride;

        var b: u32 = 0;
        while (b < blocks_per_row) : (b += 1) {
            const bp = row_ptr + b * block_stride;
            const d_all = types.fp16_to_fp32(std.mem.readInt(u16, bp[0..2], .little));
            const m_all = types.fp16_to_fp32(std.mem.readInt(u16, bp[2..4], .little));
            const scales = bp[4..];
            const qs = bp[20..];
            const x_blk = x + b * 256;

            var j: u32 = 0;
            while (j < 256) : (j += 1) {
                const sb: u32 = j / 8;
                const s = sb / 2;
                const scale = if (sb % 2 == 0)
                    d_all * @as(f32, @floatFromInt(scales[s] & 0x0F))
                else
                    d_all * @as(f32, @floatFromInt(scales[s + 8] & 0x0F));
                const min_val = if (sb % 2 == 0)
                    m_all * @as(f32, @floatFromInt(scales[s] >> 4))
                else
                    m_all * @as(f32, @floatFromInt(scales[s + 8] >> 4));
                const byte_idx: usize = j / 4;
                const bit_off: u3 = @intCast((j % 4) * 2);
                const q = (qs[byte_idx] >> bit_off) & 0x03;
                sum += (scale * @as(f32, @floatFromInt(q)) - min_val) * x_blk[j];
            }
        }
        y[m] += sum;
    }
}

test "matvec basic" {
    var W: [4 * 8]types.fp16_t = undefined;
    var x: [8]f32 = undefined;
    var y: [4]f32 = undefined;

    for (0..4) |m| {
        for (0..8) |k| {
            W[m * 8 + k] = types.fp32_to_fp16(@floatFromInt(m + 1));
        }
    }

    for (0..8) |k| {
        x[k] = @floatFromInt(k + 1);
    }

    @memset(&y, 0);

    const W_bytes: [*]const u8 = @ptrCast(&W);
    matvecFp16Fp32(8, 4, W_bytes, &x, &y, 4, 8);

    try std.testing.expectEqual(y[0], 36.0);
    try std.testing.expectEqual(y[1], 72.0);
    try std.testing.expectEqual(y[2], 108.0);
    try std.testing.expectEqual(y[3], 144.0);
}

test "matvec f32 basic" {
    var W = [_]f32{
        1, 1, 1, 1,
        2, 2, 2, 2,
    };
    var x = [_]f32{ 1, 2, 3, 4 };
    var y = [_]f32{ 0, 0 };

    const W_bytes: [*]const u8 = @ptrCast(&W);
    matvecF32Fp32(4, 2, W_bytes, &x, &y, 2, 4);

    try std.testing.expectEqual(@as(f32, 10), y[0]);
    try std.testing.expectEqual(@as(f32, 20), y[1]);
}

test "matvecQ4K matches canonical scale packing" {
    var block = std.mem.zeroes(quant.block_q4_K);
    block.d = types.fp32_to_fp16(63.0);
    block.dmin = types.fp32_to_fp16(63.0);

    for (0..4) |j| {
        block.scales[j] = @intCast(j + 1);
        block.scales[j + 4] = @intCast(j);
    }
    block.scales[0] |= 1 << 6;
    block.scales[1] |= 1 << 6;
    block.scales[2] |= 1 << 6;
    block.scales[3] |= 1 << 6;
    block.scales[8] = 0x10 | 0x02;
    block.scales[9] = 0x20 | 0x03;
    block.scales[10] = 0x30 | 0x04;
    block.scales[11] = 0x40 | 0x05;
    @memset(&block.qs, 0x55);

    var x: [256]f32 = undefined;
    for (0..x.len) |i| x[i] = @as(f32, @floatFromInt((i % 11) + 1)) * 0.01;

    var y = [_]f32{0};
    matvecQ4K(std.mem.asBytes(&block).ptr, &x, &y, 1, 256);

    const d_f32 = types.fp16_to_fp32(block.d);
    const dmin_f32 = types.fp16_to_fp32(block.dmin);
    var expected: f32 = 0;
    for (0..4) |blk| {
        const js = blk * 2;
        const sc0: u8 = if (js < 4)
            block.scales[js] & 63
        else
            (block.scales[js + 4] & 0x0F) | ((block.scales[js - 4] >> 6) << 4);
        const mn0: u8 = if (js < 4)
            block.scales[js + 4] & 63
        else
            (block.scales[js + 4] >> 4) | ((block.scales[js] >> 6) << 4);
        const sc1: u8 = if (js + 1 < 4)
            block.scales[js + 1] & 63
        else
            (block.scales[js + 1 + 4] & 0x0F) | ((block.scales[js + 1 - 4] >> 6) << 4);
        const mn1: u8 = if (js + 1 < 4)
            block.scales[js + 1 + 4] & 63
        else
            (block.scales[js + 1 + 4] >> 4) | ((block.scales[js + 1] >> 6) << 4);

        const rs0 = d_f32 * @as(f32, @floatFromInt(sc0));
        const rm0 = dmin_f32 * @as(f32, @floatFromInt(mn0));
        const rs1 = d_f32 * @as(f32, @floatFromInt(sc1));
        const rm1 = dmin_f32 * @as(f32, @floatFromInt(mn1));
        const qoff = blk * 32;
        const woff = blk * 64;

        for (0..16) |j| {
            expected += (rs0 * @as(f32, @floatFromInt(block.qs[qoff + j] & 0x0F)) - rm0) * x[woff + j];
            expected += (rs1 * @as(f32, @floatFromInt(block.qs[qoff + 16 + j] & 0x0F)) - rm1) * x[woff + 16 + j];
            expected += (rs0 * @as(f32, @floatFromInt(block.qs[qoff + j] >> 4)) - rm0) * x[woff + 32 + j];
            expected += (rs1 * @as(f32, @floatFromInt(block.qs[qoff + 16 + j] >> 4)) - rm1) * x[woff + 48 + j];
        }
    }

    try std.testing.expectApproxEqAbs(expected, y[0], 0.01);
}
